Aead retry token (#59)
quicwg/load-balancers: draft-ietf-quic-load-balancers.md

Both server and service must have access to Universal time, though tight
synchronization is unnecessary.

The tokens are protected using AES128-GCM AEAD, as explained in
{{token-protection-with-aead}}.
All tokens, generated by either the server or retry service, MUST use the
following format, which includes:

> - A 96 bit token sequence number transmitted in clear text, but

Reviewer: ```suggestion
- A 96 bit unique token number transmitted in clear text, but
```

  protected as part of the AEAD associated data.
- An 8 bit token key identifier.
- A token body, encoding the Original Destination Connection ID,
  the Retry Source Connection ID, and the Timestamp, optionally
  followed by server specific Opaque Data.

The token protection uses an 128 bit representation of the
source IP address from the triggering Initial packet.
The client IP address is 16 octets. If an IPv4 address, the last 12 octets are
zeroes. If there is a Network Address Translator (NAT) in the server
infrastructure that changes the client IP, the Retry Service MUST either be
positioned behind the NAT, or the NAT must have the token key to rewrite the
Retry token accordingly.

> The 96 but token sequence number is set to a random value

Reviewer: ```suggestion
The 96 bit unique token number is set to a random value
```

using a cryptography grade random number
generator. 

The token key identifier and the corresponding

> AEAD key and AEAD IV are provisioned by the server. The server also

Reviewer: ```suggestion
AEAD key and AEAD IV are provisioned by the configuration agent. 
```

> provisions the maximum duration of a token encrypted with that key.

Reviewer: ```suggestion
```

Both server and service must have access to Universal time, though tight
synchronization is unnecessary.

The tokens are protected using AES128-GCM AEAD, as explained in
{{token-protection-with-aead}}.
All tokens, generated by either the server or retry service, MUST use the
following format, which includes:

- The client IP address, serialized and used as associated data when
  computing the AEAD encryption.
- A token sequence number that must be used only once with a given
  key. The token sequence number is transmitted in clear text, but
  is protected as part of the AEAD associated data.
- A token body, encoding the Original Destination Connection ID,
  the Retry Source Connection ID, and the Timestamp, optionally
  followed by server specific Opaque Data.

DISCUSS: shall we specify how the token sequence is formatted? There
are two main possibilities, a large random number that is statistically

> unique, or the combination of a server ID and a sequence number unique

Reviewer: I like the ID of prepending the server ID to the sequence number and assigning the Retry Service one or more "server IDs" that keep the sequence number spaces separate, but one problem is that the server ID can be as much as 18 Bytes.

I see three ways forward:
- assign each server a second server ID for these purposes, (blecch)
- just use random numbers
- eliminate the (already very weak) ability of Shared State Retry Services to validate incoming tokens -- they only generate them. Then, each server can do whatever it wants with its self-generated tokens, which should easily be distinguishable by token length or whatever, and the service can use the entire sequence number space without fear of collision.

I'm leaning towards the third.

Author: Birthday paradox says with a 64 bit random field you will expect collisions after 2^32 tokens -- and that you have to change the key sooner than that if you want to minimize the risk. If you want the risk of collision below 10E-5 (five nines...) then you should not use the key more than 19 million times.

I should be technically possible to make the identifier 12 bytes long -- that's the normal size of the IV for AES GCM. The same reasoning says that the key should not be use more that 4.0 E+14 times -- 400 trillion times. That may be the way to go. Better than trying to squeeze an 18 bytes ID there.

Or we would have to use something a bit more exotic than AES GCM, but I would rather not go there.

Reviewer: One way to manage the key rotation would be to actually configure both entities with a secret and make the retry service and server derive the key from that, so that it can be rotated using some signal that the retry service puts in the token?

Reviewer: Or we could just do my third option...

to hat server.

DISCUSS: some implementations of QUIC encode in the token the Initial Packet

> Number used by the client, in order to verify that the client sends the

Reviewer: Can you file a separate issue for that? It applies to both types of Retry Service and we should address it consistently for both.

I don't feel strongly one way or the other. I believe the point of that check is so that *some* servers enforce compliance to encourage good behavior, and I don't think Retry Services need to be part of that effort.

Author: The PN s only used with the RETRY TOKEN format, which is server specific. So it fits naturally in the server specific part.

Reviewer: Yes, but a non-shared-state retry service might be required to include the initial packet number so that the server can do the validation.

That said, making the retry service do initial header decryption is a terrible idea. So let's rule out having any type of retry service encode the packet number.

retried Initial with a PN larger that the triggering Initial. Should we specify
that here?

The token sequence number is encoded as a 64 bit binary field:
~~~~~
0                   1                   2                   3
0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                 Token sequence number (64)                    +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~~~

> The token body is encoded as follow:

Reviewer: s/follow/follows

Author: OK, will fix.

Both server and service must have access to Universal time, though tight
synchronization is unnecessary.

The tokens are protected using AES128-GCM AEAD, as explained in
{{token-protection-with-aead}}.
All tokens, generated by either the server or retry service, MUST use the
following format, which includes:

> - A 32 bit server ID, transmitted in clear text and protected as part

Reviewer: We agreed over the phone not to have a "controlled" configuration, so we'll just consolidate these two fields into a 96-bit nonce.

implementations will encode the Initial Packet
Number as part of the opaque data.

### Token Protection with AEAD {#token-protection-with-aead}

On the wire, the
token is presented as:
~~~~~
0                   1                   2                   3
0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                                                               +
|                 Token sequence number (96)                    |
+                                                               +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
| Key seq.(8)   |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
| AEAD Encrypted Token (variable)                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
| AEAD Checksum (variable, function of encryption algorithm)    |                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~

> The tokens are protected using AES128-GCM as follow:

Reviewer: ```suggestion
The tokens are protected using AES128-GCM as follows:
```

* The token key and IV are retrieved using the Key Sequence.

> * The nonce, N, is formed by combining the IV with the 32 bit server ID 

Reviewer: ```suggestion
*
```

> and the 64 bit token

Reviewer: ```suggestion
```

> sequence number, producing a 96 bit unique token number.

Reviewer: ```suggestion

```

The 96 bits of the unique token number are left-padded

> with zeros to the size of the IV. The exclusive OR of the padded token sequence

Reviewer: ```suggestion
with zeros to the size of the IV. The exclusive OR of the padded unique token
```

0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                                                               +
|                                                               |
+                        IP Address (128)                       +
|                                                               |
+                                                               +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                     Server ID (32)                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                 Token sequence number (64)                    +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
| Key seq.(8)   |
+-+-+-+-+-+-+-+-+
~~~~~
* The input plaintext for the AEAD is the token body. The output ciphertext

> of the AEAD is transmitted in place of token body.

Reviewer: ```suggestion
of the AEAD is transmitted in place of the token body.
```

distinguish between retry tokens (which trigger use of certain transport
parameters) and NEW_TOKEN frame tokens.

Some implementations of QUIC encode in the token the Initial Packet
Number used by the client, in order to verify that the client sends the
retried Initial with a PN larger that the triggering Initial. Such
implementations will encode the Initial Packet
Number as part of the opaque data.

> DISCUSS: we have a problem of intervals. We can either specify that the timestamp

Reviewer: If it's the expiration time, then each token generator can specify its own token expiration time; it it's the start time, than the token consumer is always in control of the expiration. The difference is not huge. Consumer decisions happen later, and therefore have (slightly) more information. This is related to your DISCUSS point at Line 871; if the Retry service isn't checking timestamps, than it seems easier to configure if only the servers need to worry about the proper expiration interval.

I don't feel strongly either way.

Reviewer: We talked and realized that as the NEW_TOKEN is also following this format, it is far better for the token to encode the expiration time so as not to force the Service to implement multiple policies and disambiguate the two types.

|                     Server ID (32)                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                 Token sequence number (64)                    +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
| Key seq.(8)   |
+-+-+-+-+-+-+-+-+
~~~~~
* The input plaintext for the AEAD is the token body. The output ciphertext
of the AEAD is transmitted in place of token body.
* The AEAD Checksum is computed as part of the AEAD encryption process,
and is verified during decryption.

### Configuration Agent Actions

The configuration agent generates and distributes a "token key", a "token IV",
a key sequence,
and a list of
QUIC versions the service supports. It must also inform the service if a NAT
lies between the service and the servers.

The configuration also specifies whether the unique token number is
generated using the "random" or the "controlled" methods. If the controlled
method is used, the configuration agent provisions each server with a

> unique 32 bit server ID.

Reviewer: The Retry Service also needs a "generator ID". Arguably, it would make sense to have more than one (it presumably will generate tokens in much higher volume).

Reviewer: Disregard, no server IDs anymore

### Service Requirements

When in active mode, the service MUST generate Retry tokens with the format
described above when it receives a client Initial packet with no token.

In active mode, the service SHOULD decrypt the incoming
tokens. The service SHOULD drop packets with unknown key sequence,
or an AEAD checksum that does not match the expected value.
(By construction, the AEAD checksum will only match is the client
IP Address also matches.)
It SHOULD forward packets that do, regardless of the other fields.

> DISCUSS: should we say something about the time stamp?

Reviewer: Why wouldn't it? As there are multiple token generators in this model, the time thresholds are going to have be forgiving about clock skew anyway.

Reviewer: Result of call: yes, the service should look at the timestamp (with allowances for clock skew)

~~~~~
0                   1                   2                   3
0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                                                               +
|                                                               |
+                        IP Address (128)                       +
|                                                               |
+                                                               +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                                                               |
+                        Timestamp (64)                         +
|                                                               |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
~~~~~
* The input plaintext, for the AEAD is the token body.The output ciphertext
of the AEAD is transmitted in place of token body. On the wire, the

> token is presented as:

Reviewer: Should this make clear that the authentication tag is appended to the end of the token (assuming it is)?

Author: Yes it is. I copied the text from the QUIC-TLS draft, the tag is considered part of the encrypted data.

Reviewer: With the "opaque" data and so on it would be helpful to show exactly where the tag goes, so as to leave no ambiguity.
